File: kombi-lib/src/main/java/com/sgnatiuk/cartesian/CombinationMask.java
package com.sgnatiuk.cartesian;

import java.util.Iterator;

class CombinationMask implements Iterator<int[]> {

    private final int[] bases;
    private final int[] mask;
    private boolean hasNext = true;

    public CombinationMask(int[] bases) {
        if (bases.length == 0) {
            throw new IllegalArgumentException("Expected non empty array");
        }
        this.bases = bases;
        this.mask = new int[bases.length];
    }

    @Override
    public boolean hasNext() {
        return hasNext || (hasNext = !increment());
    }

    @Override
    public int[] next() {
        hasNext = false;
        return mask;
    }

    /**
     * Computes next encoded combination
     *
     * @return true if encoded was overflowed, otherwise false
     */
    private boolean increment() {
        int index = 0;
        do {
            int incrementedCell = mask[index] + 1;
            if (incrementedCell == bases[index]) {
                mask[index++] = 0;
            } else {
                mask[index] = incrementedCell;
                return false;
            }
        } while (index < bases.length);

        return true;
    }
}

File: src/main/kotlin/com/sgnatiuk/encodable/encoders/CombinationMask.kt
package com.sgnatiuk.encodable.encoders

import java.util.*

internal class CombinationMask(
        private val bases: IntArray
) : Iterator<IntArray> {

    private val encoded: IntArray = IntArray(bases.size)
    private var hasNextElement : Boolean = true


    override fun hasNext(): Boolean {
        if(hasNextElement) return true

        hasNextElement = !encoded.increment()
        return hasNextElement
    }

    override fun next() = encoded.apply {
        hasNextElement = false
    }

    /**
     * Computes next encoded combination
     * @return true if encoded was overflowed, otherwise false
     */
    private fun IntArray.increment(): Boolean {
        var extra = 1
        var index = this.size - 1
        do {
            val temp = this[index] + extra
            this[index] = temp % bases[index]
            extra = temp / bases[index]
            index--
        } while (extra != 0 && index >= 0)
        return index < 0 && extra > 0
    }

    private fun copy(arr: IntArray): IntArray {
        return Arrays.copyOf(arr, arr.size)
    }
}